gplugins/lumerical/eme.py

```diff
@@ -0,0 +1,1063 @@
+        subcell_methods = []
+        for method in ss.group_subcell_methods:
+            if method == "CVCS":
+                subcell_methods.append(1)
+            else:
+                subcell_methods.append(0)
+        s.set("subcell method", np.array(subcell_methods))
+
+        s.set("number of modes for all cell groups", ss.num_modes)
+        s.set("energy conservation", ss.energy_conservation)
+
+        s.set("define y mesh by", "maximum mesh step")
+        s.set("define z mesh by", "maximum mesh step")
+
+        s.set("dy", ss.wavelength / ss.mesh_cells_per_wavelength * um)
+        s.set("dz", ss.wavelength / ss.mesh_cells_per_wavelength * um)
+
+        s.set("y min bc", ss.ymin_boundary)
+        s.set("y max bc", ss.ymax_boundary)
+        s.set("z min bc", ss.zmin_boundary)
+        s.set("z max bc", ss.zmax_boundary)
+
+        s.set("pml layers", ss.pml_layers)
+
+        s.save(str(dirpath / f"{component.name}.lms"))
+
+        if run_overall_convergence:
+            if not hide:
+                logger.info("Running overall convergence")
+            self.update_overall_convergence(plot=not hide)
+        else:
+            if run_mesh_convergence:
+                if not hide:
+                    logger.info("Running mesh convergence.")
+                self.mesh_convergence_data = self.update_mesh_convergence(plot=not hide)
+
+            if run_cell_convergence:
+                if not hide:
+                    logger.info("Running cell convergence.")
+                self.cell_convergence_data = self.update_cell_convergence(plot=not hide)
+
+        if run_mode_convergence:
+            if not hide:
+                logger.info("Running mode convergence.")
+            self.mode_convergence_data = self.update_mode_convergence(plot=not hide)
+
+        if not hide:
+            plt.show()
+
+    def update_mesh_convergence(self, plot: bool = False) -> pd.DataFrame:
+        """
+        Update simulation based on mesh convergence testing. Updates both Lumerical session and simulation settings.
+
+        Parameters:
+            plot: Plot and save convergence results
+
+        Returns:
+            Convergence results
+        """
+
+        s = self.session
+        cs = self.convergence_settings
+        ss = self.simulation_settings
+
+        s21 = []
+        s11 = []
+        mesh_cells_per_wavl = []
+        converged = False
+        while not converged:
+            s.switchtolayout()
+            s.set("dy", ss.wavelength / ss.mesh_cells_per_wavelength * um)
+            s.set("dz", ss.wavelength / ss.mesh_cells_per_wavelength * um)
+            # Get sparams and refine mesh
+            s.run()
+            s.emepropagate()
+            S = s.getresult("EME", "user s matrix")
+            s11.append(abs(S[0, 0]) ** 2)
+            s21.append(abs(S[1, 0]) ** 2)
+            mesh_cells_per_wavl.append(ss.mesh_cells_per_wavelength)
+
+            ss.mesh_cells_per_wavelength += 1
```

Comment:
should probably have a comment here

Reply:
Done in latest
https://github.com/seanlam97/gplugins/blob/a135b4bfcfbe6db972d8e3297c5291c593369a0d/gplugins/lumerical/eme.py